### ISSUE TITLE
`audit`: check availability of URLs and HTTPS versions

### ISSUE THREAD

**Issue by vitorgalvao** [Homebrew/homebrew-cask#25403]
Just thought this might be a good addition to `audit`. Checking if URLs are accessible and HTTPS versions are available. Since we can’t take only 200 HTTP codes to mean success (since some servers have weird replies), I’m defining accessible as everything with 2-- and 3-- codes.

Proposal. When auditing a cask check all `url`, `appcast`, and `homepage` stanzas, and for each of them:
- If http:
  - Try the exact same url but with http**s** instead, following all redirects.
    - If the last location is http**s**, check if it is accessible.
      - If yes, fail audit and ask the user to change the url to http**s**.
      - If no, revert to the user provided http url and see if it is accessible.
        - If yes, pass audit.
        - If no, fail audit and ask the user to verify if the url is correct.
- If http**s**:
  - Try to access the url, following all redirects.
    - If the last location continues to be http**s**, check if accessible.
      - If yes, pass audit.
      - If no, try the same url with http instead and see if it is accessible.
        - If yes, fail audit and warn the user that although the url is accessible, it must be changed to http
        - If no, fail audit and ask the user to verify if the url is correct.

Note that if a user provides an HTTPS URL and it redirects to an HTTP one, `audit` will always fail (with different messages), regardless of it being accessible or inaccessible. This is in line with our long-standing position of not inducing users in error.

**Comment by adidalal** [comment]
Seems reasonable, and is a well-scoped feature

**Comment by miccal** [comment]
@victorpopkov has recently created a [cask-homepage](https://github.com/victorpopkov/cask-scripts/blob/master/doc/cask-homepage.md) script that is able to check if the `homepage` stanza allows `https`, and submitted a [PR](https://github.com/caskroom/homebrew-cask/pull/26759) to fix them.

Perhaps some ideas from his script could be used for this?

**Comment by vitorgalvao** [comment]
@miccal That (and similar scripts) work only after the fact. This is about making it work directly in `audit`.

**Comment by miccal** [comment]
True, but doesn't `audit` work after the fact also? I mean, a Cask file is created first, then `audit` is run on the file?

**Comment by vitorgalvao** [comment]
But it’s done *before* submission *and* checked by Travis, which is what matters.

**Comment by miccal** [comment]
I think I get what your saying - the cask-homepage script will _only_ work on a Cask file that is already a part of Homebrew-Cask, while `audit` works on _any_ Cask file, merged or not?

**Comment by vitorgalvao** [comment]
Yes, that’s exactly it. Plus, since `audit` is part of the core and it *has to run*, it won’t be possible to not perform this check.

**Comment by miccal** [comment]
Got it, thanks!

**Comment by Git-Jiro** [comment]
I will have a look at implementing this feature.

**Comment by commitay** [comment]
> https://travis-ci.org/caskroom/homebrew-cask/builds/312042252#L753
```
audit for tankix: failed
 - The URL http://static.tankix.com/app/StandaloneOSXIntel64/master-9108/TankiX.dmg should use HTTPS rather than HTTP
Error: audit failed for 1 cask: tankix
```

**Comment by amyspark** [comment]
Reviewing old issues here - was this fixed by Homebrew/brew#3318?

**Comment by amyspark** [comment]
@vitorgalvao or @commitay - can you outline what's still missing?
